### ISSUE TITLE
assert_smaller for `(Sum rest)` is structurally smaller than `(Sum ((pi, P) :: rest))`

### ISSUE THREAD

**Issue by Jake-Gillberg** [dckc/togl#3]
https://github.com/dckc/togl/blob/664731697c691137bb13a3b125ed405ced568b77/picalc2.lidr#L120

I got coq to recognize this by mutually defining a recursive function for fn on Sums and Processes, I think a similar thing would work in idris, but I haven't gotten there yet.

In coq:
https://github.com/Jake-Gillberg/Pi-Proofs/blob/3901cd81165787964926151708c23b159cb3364d/pi.v#L23

P.S. Hi :)

**Comment by dckc** [comment]
yes, there should be a way to use a measure... I can't remember how it works any more, exactly, but I collaborated on https://github.com/idris-lang/Idris-dev/blob/master/libs/prelude/Prelude/WellFounded.idr which uses it.
